### PR TITLE
Sparse kernel development for GPU. Part 2

### PR DESCRIPTION
**Pull request by nychiang**
This PR includes the following changes:
1. add hiopMatrixRajaSparseTriplet::copySubMatrixFromTrans and unit test
2. add hiopMatrixRajaSparseTriplet::copySubMatrixFrom and unit test

### CODE REVIEW

**Review by ashermancinelli — APPROVED** [review]

File: tests/LinAlg/matrixTestsSparseTriplet.cpp [hunk 1]
```diff
@@ -226,7 +226,7 @@ int MatrixTestsSparseTriplet::verifyAnswer(
     {
       if (!isEqual(getLocalElement(A, i, j), expect(i, j)))
       {
-        //printf("(%d, %d) failed. %f != %f.\n", i, j, getLocalElement(A, i, j), expect(i, j));
+        printf("(%d, %d) failed. %f != %f.\n", i, j, getLocalElement(A, i, j), expect(i, j));
```
Comment (cnpetra) [inline tests/LinAlg/matrixTestsSparseTriplet.cpp:229]: do we need this in here?

File: src/LinAlg/hiopMatrixSparse.hpp [hunk 2]
```diff
@@ -180,12 +187,46 @@ class hiopMatrixSparse : public hiopMatrix
    * @pre 'dest_nnz_st' + the number of non-zeros in the copied the rows must be less or equal to this->numOfNumbers()
    * @pre User must know the nonzero pattern of src and dest matrices. Assume non-zero patterns of these two wont change, and 'src_gen' is a submatrix of 'this'
    * @pre Otherwise, this function may replace the non-zero values and nonzero patterns for the undesired elements.
+   * @pre: this function does NOT preserve the sorted row/col indices. USE WITH CAUTION!
    */
   virtual void copyRowsBlockFrom(const hiopMatrix& src_gen,
```
Comment (cnpetra) [inline src/LinAlg/hiopMatrixSparse.hpp:192]: alignment and codestyle  are off

File: src/LinAlg/hiopMatrixRajaSparseTriplet.cpp [hunk 3]
```diff
@@ -1417,6 +1413,89 @@ void hiopMatrixRajaSparseTriplet::set_Jac_FR(const hiopMatrixSparse& Jac_c,
...
+        index_type dest_k = dest_nnz_st + src_k;
+        iRow[dest_k] = dest_row_st + src_iRow[src_k];
+        jCol[dest_k] = dest_col_st + src_jCol[src_k];
+        values[dest_k] = src_val[src_k];
+      }
+    }
+  );
+}
+
+/// @brief copy a submatix from a transpose of another matrix.
```
Comment (cnpetra) [inline src/LinAlg/hiopMatrixRajaSparseTriplet.cpp:1457]: typo. should be "submatrix"

File: src/LinAlg/hiopMatrixRajaSparseTriplet.cpp [hunk 4]
```diff
@@ -1417,6 +1413,89 @@ void hiopMatrixRajaSparseTriplet::set_Jac_FR(const hiopMatrixSparse& Jac_c,
   copyFromDev();
 }
 
+/// @brief copy a submatix from another matrix.
```
Comment (cnpetra) [inline src/LinAlg/hiopMatrixRajaSparseTriplet.cpp:1416]: typo. "submatrix"

**Comment by nychiang** [comment]
@cnpetra all the small comments have been addressed

**Review by cnpetra — APPROVED** [review]

**Pull request merged**